### DISCUSSION TITLE
Add pre-commit as an option with built-in opinions

### DISCUSSION THREAD

**Discussion by ericmjl** [drivendataorg/cookiecutter-data-science#382] [Requested tools]
I would like to propose adding in an opinionated `.pre-commit-config.yaml`, to enable pre-commit hooks in the project. Doing so would help push automated code checks directly to the local repo clone, and can help with ensuring code quality.

In terms of whether or not this should be enabled by default, I would lean towards not enabling it by default, as the use of pre-commit is not commonly taught in introductory `git` courses. However, I am quite sure that seasoned `git` users will appreciate its availability in the repository nonetheless.

Before I get working on it, was wondering what you all might think about this contribution? As this would be my first contribution to the repo, I am happy to follow your guidance on what else should be added.

**Comment by sobolevn** [comment]
I would also love to see this feature! Consider adding [`wemake-python-styleguide`](https://github.com/wemake-services/wemake-python-styleguide) as a part of the `pre-commit` pipeline

**Comment by fgriberi** [comment]
Hi guys, can I take this issue? What I have in mind is a pre-commit with the following plugin:
* flake8
* isort
* wemake-python-styleguide
* black

please, let me know your thoughts

**Comment by sobolevn** [comment]
@fgriberi feel free to ping me for review / help 👍

**Comment by pjbull** [comment]
I like the idea of this. A few additional thoughts:

 - We've done a [`register-hook` make command](https://github.com/drivendataorg/deon/blob/master/Makefile#L1-L6) for easy installation in the past. I think by default that command should be optional (but, our team often makes it a dependency of requirements).
 - If the [precommit hook runs make commands](https://github.com/drivendataorg/deon/blob/master/.precommithook#L3), then committing updated make commands updates the hook for everyone, which is convenient.
 - A few opinions: `wemake-python-styleguide` is too opinionated for a default. `flake8` is a definite for me, `black` is probably, `isort` I could take or leave but don't feel strongly.

Also, it is worth noting that the current project template is supposed to be VCS agnositic, so it almost makes more sense once #162 is in. (Which we are pushing right now to get CI working for, if anyone wants to look at #205 where @jayqi was exploring ways to test virtualenvwrapper on Windows).

**Comment by fgriberi** [comment]
@pjbull I've been thinking about using [pre-commit](https://pre-commit.com/) instead of a bash script. The integration with the mentioned tools is awesome. What do you think?

**Comment by pjbull** [comment]
I didn't find the increase in complexity to be worth it for our team, and I think that it sacrifices the flexibility that a project template like this should have.

**Comment by smb-h** [comment]
Hi everyone,

I've created a pull request (#323) that addresses this issue. It adds pre-commit hooks for code formatting and linting to enhance code quality and consistency in the project. The pre-commit hooks include tools such as `pre-commit-hooks`, `mirrors-prettier`, `black`, `isort`, and `flake8`.

You can review the pull request here: [Pull Request #323](https://github.com/drivendata/cookiecutter-data-science/pull/323)

Your feedback and suggestions are greatly appreciated. Please take a look at the changes, and if everything looks good, your approval would be fantastic. If you have any comments or questions, feel free to share them in the PR discussion.

Let's work together to improve the code quality of this project! Thank you for your attention to this matter.
